File: src/bst/Treap.java
package bst;

import java.util.Random;

public class Treap {

// private:

    private static final Random GENERATOR = new Random();

    private final Treap left;
    private final Treap right;
    private final int priority;
    private final int key;

    private Treap(int x, int y, Treap left, Treap right) {
        this.key = x;
        this.priority = y;
        this.left = left;
        this.right = right;
    }

    private Treap cloneNode(Treap left, Treap right) {
        return new Treap(this.key, this.priority, left, right);
    }


// public:

    public static Treap create(int key) {
        int prior = GENERATOR.nextInt();
        return new Treap(key, prior, null, null);
    }

    public static Treap merge(Treap l, Treap r) {
        if (l == null) {
            return r;
        }
        if (r == null) {
            return l;
        }
        if (l.priority > r.priority) {
            return l.cloneNode(l.left, merge(l.right, r));
        } else {
            return r.cloneNode(merge(l, r.left), r.right);
        }
    }

    public Treap[] split(int pivot) {
        if (this.key < pivot) {
            Treap[] tuple = (this.right == null)
                    ? new Treap[] {null, null}
                    : this.right.split(pivot);
            Treap l1 = this.cloneNode(this.left, tuple[0]);
            return new Treap[] {l1, tuple[1]};
        } else if (this.key > pivot) {
            Treap[] tuple = (this.left == null)
                    ? new Treap[] {null, null}
                    : this.left.split(pivot);
            Treap r1 = this.cloneNode(tuple[1], this.right);
            return new Treap[] {tuple[0], r1};
        } else {
            return new Treap[] {this.left, this.right};
        }
    }

    public Treap insert(int key) {
        Treap[] tuple = this.split(key);
        return merge(tuple[0], merge(create(key), tuple[1]));
    }

    public String toString() {
        String ll = (this.left == null) ? "" : this.left.toString();
        String rr = (this.right == null) ? "" : right.toString();
        return "(" + ll + " " + this.key + "/" + this.priority + " " + rr + ")";
    }

}

File: src/Treap.scala
import java.util.Random

object Treap {
  private val GENERATOR: Random = new Random

  def create(key: Int): Treap = {
    val prior: Int = GENERATOR.nextInt
    new Treap(key, prior, null, null)
  }

  def merge(l: Treap, r: Treap): Treap = {
    if (l == null) {
      return r
    }
    if (r == null) {
      return l
    }
    if (l.priority > r.priority) {
      l.cloneNode(l.left, merge(l.right, r))
    }
    else {
      r.cloneNode(merge(l, r.left), r.right)
    }
  }
}

case class Treap(key: Int, priority: Int, left: Treap = null, right: Treap = null) {

  private def cloneNode(left: Treap, right: Treap): Treap = {
    new Treap(this.key, this.priority, left, right)
  }

  def split(pivot: Int): (Treap, Treap) = {
    if (this.key < pivot) {
      val tuple = if (this.right == null) {
        (null, null)
      } else {
        this.right.split(pivot)
      }
      val l1: Treap = this.cloneNode(this.left, tuple._1)
      (l1, tuple._2)
    }
    else if (this.key > pivot) {
      val tuple = if (this.left == null) {
        (null, null)
      } else {
        this.left.split(pivot)
      }
      val r1: Treap = this.cloneNode(tuple._2, this.right)
      (tuple._1, r1)
    }
    else {
      (this.left, this.right)
    }
  }

  def insert(key: Int): Treap = {
    val tuple: (Treap, Treap) = this.split(key)
    Treap.merge(tuple._1, Treap.merge(Treap.create(key), tuple._2))
  }

  override def toString: String = {
    val ll: String = if (this.left == null) "" else this.left.toString
    val rr: String = if (this.right == null) "" else right.toString
    "(" + ll + " " + this.key + "/" + this.priority + " " + rr + ")"
  }

}
